src/devices/virtio-blk.c

```diff
@@ -0,0 +1,442 @@
+                                   uint32_t *plen)
+{
+    /* A full virtio_blk_req is represented by 3 descriptors, where
+     * the first descriptor contains:
+     *   le32 type
+     *   le32 reserved
+     *   le64 sector
+     * the second descriptor contains:
+     *   u8 data[][512]
+     * the third descriptor contains:
+     *   u8 status
+     */
+    struct virtq_desc vq_desc[3];
+
+    /* Collect the descriptors */
+    for (int i = 0; i < 3; i++) {
+        /* The size of the `struct virtq_desc` is 4 words */
+        const uint32_t *desc = &vblk->ram[queue->QueueDesc + desc_idx * 4];
+
+        /* Retrieve the fields of current descriptor */
+        vq_desc[i].addr = desc[0];
+        vq_desc[i].len = desc[2];
+        vq_desc[i].flags = desc[3];
+        desc_idx = desc[3] >> 16; /* vq_desc[desc_cnt].next */
+    }
+
+    /* The next flag for the first and second descriptors should be set,
+     * whereas for the third descriptor is should not be set
+     */
+    if (!(vq_desc[0].flags & VIRTIO_DESC_F_NEXT) ||
+        !(vq_desc[1].flags & VIRTIO_DESC_F_NEXT) ||
+        (vq_desc[2].flags & VIRTIO_DESC_F_NEXT)) {
+        /* since the descriptor list is abnormal, we don't write the status
+         * back here */
+        virtio_blk_set_fail(vblk);
+        return -1;
+    }
+
+    /* Process the header */
+    const struct vblk_req_header *header =
+        (struct vblk_req_header *) ((uintptr_t) vblk->ram + vq_desc[0].addr);
+    uint32_t type = header->type;
+    uint64_t sector = header->sector;
+    uint8_t *status = (uint8_t *) ((uintptr_t) vblk->ram + vq_desc[2].addr);
+
+    /* Check sector index is valid */
+    if (sector > (PRIV(vblk)->capacity - 1)) {
+        *status = VIRTIO_BLK_S_IOERR;
+        return -1;
+    }
+
+    /* Process the data */
+    switch (type) {
+    case VIRTIO_BLK_T_IN:
+        virtio_blk_read_handler(vblk, sector, vq_desc[1].addr, vq_desc[1].len);
+        break;
+    case VIRTIO_BLK_T_OUT:
+        virtio_blk_write_handler(vblk, sector, vq_desc[1].addr, vq_desc[1].len);
+        break;
+    default:
+        fprintf(stderr, "unsupported virtio-blk operation!\n");
+        *status = VIRTIO_BLK_S_UNSUPP;
+        return -1;
+    }
+
+    /* Return the device status */
+    *status = VIRTIO_BLK_S_OK;
+    *plen = vq_desc[1].len;
+
+    return 0;
+}
+
+static void virtio_queue_notify_handler(virtio_blk_state_t *vblk, int index)
+{
+    uint32_t *ram = vblk->ram;
+    virtio_blk_queue_t *queue = &vblk->queues[index];
+    if (vblk->Status & VIRTIO_STATUS__DEVICE_NEEDS_RESET)
+        return;
+
+    if (!((vblk->Status & VIRTIO_STATUS__DRIVER_OK) && queue->ready))
+        return virtio_blk_set_fail(vblk);
+
+    /* Check for new buffers */
+    uint16_t new_avail = ram[queue->QueueAvail] >> 16;
+    if (new_avail - queue->last_avail > (uint16_t) queue->QueueNum)
+        return (fprintf(stderr, "size check fail\n"),
+                virtio_blk_set_fail(vblk));
```

Comment:
Split into two statements. It would be better for static analyzer to locate the source lines.